python/GafferScene/RenderPassTypeAdaptor.py

```diff
@@ -0,0 +1,152 @@
+
+		self["__ContextQuery"] = Gaffer.ContextQuery()
+		self["__ContextQuery"].addQuery( Gaffer.StringPlug() )
+		self["__ContextQuery"]["queries"][0]["name"].setValue( "renderPass" )
+
+		self["__NameSwitch"] = Gaffer.NameSwitch()
+		self["__NameSwitch"].setup( GafferScene.ScenePlug() )
+		self["__NameSwitch"]["in"]["in0"]["value"].setInput( self["in"] )
+
+		self["__AutoTypeExpression"] = Gaffer.Expression()
+		self["__AutoTypeExpression"].setExpression(
+			inspect.cleandoc(
+				"""
+				import GafferScene
+
+				renderPassType = parent["__OptionQuery"]["out"]["out0"]["value"]
+				autoTypeFunction = GafferScene.RenderPassTypeAdaptor.autoTypeFunction()
+				if renderPassType == "auto" and callable( autoTypeFunction ) :
+					renderPassName = parent["__ContextQuery"]["out"]["out0"]["value"]
+					renderPassType = autoTypeFunction( renderPassName )
+
+				parent["__NameSwitch"]["selector"] = renderPassType
+				"""
+			)
+		)
+
+		global __renderPassTypeProcessors
+		for type in self.registeredTypeNames() :
+			outputPlug = self["in"]
+
+			for name, p in __renderPassTypeProcessors[type].items() :
+				processor = p()
+				processor["in"].setInput( outputPlug )
+				self.addChild( processor )
+				processor.setName( f"{type}_{name}" )
+				if processor.getChild( "client" ) :
+					processor["client"].setInput( self["client"] )
+				if processor.getChild( "renderer" ) :
+					processor["renderer"].setInput( self["renderer"] )
+				outputPlug = processor["out"]
+
+			plug = self["__NameSwitch"]["in"].next()
+			plug["value"].setInput( outputPlug )
+			plug["name"].setValue( type )
+
+		self["out"].setInput( self["__NameSwitch"]["out"]["value"] )
+		self['out'].setFlags( Gaffer.Plug.Flags.Serialisable, False )
+
+	global __renderPassTypeProcessors
+	__renderPassTypeProcessors = collections.defaultdict( dict )
+
+	@staticmethod
+	def registerTypeProcessor( type, name, f ) :
+
+		global __renderPassTypeProcessors
+		__renderPassTypeProcessors[type][name] = f
+
+	@staticmethod
+	def registeredTypeNames() :
+
+		global __renderPassTypeProcessors
+		return __renderPassTypeProcessors.keys()
+
+	@staticmethod
+	def registeredTypeProcessors( type ) :
+
+		global __renderPassTypeProcessors
+		return __renderPassTypeProcessors[type].keys()
+
+	@staticmethod
+	def deregisterTypeProcessor( type, name ) :
+
+		global __renderPassTypeProcessors
+		if type in __renderPassTypeProcessors.keys() and name in __renderPassTypeProcessors[type].keys() :
+			del __renderPassTypeProcessors[type][name]
+
+			if not __renderPassTypeProcessors[type].keys() :
+				del __renderPassTypeProcessors[type]
+
+	global __renderPassAutoTypeFunction
+	__renderPassAutoTypeFunction = None
+
+	@staticmethod
+	def registerAutoTypeFunction( f ) :
+
+		global __renderPassAutoTypeFunction
+		__renderPassAutoTypeFunction = f
+
+	@staticmethod
+	def autoTypeFunction() :
+
+		global __renderPassAutoTypeFunction
+		return __renderPassAutoTypeFunction
```

Comment:
Sounds good to me, added `resolvedType` in fbb96c5efc3634cb3fe23335f74d70a719e8dc67.

Reply:
I think it might be good to have some sort of handling for when `type == "auto"` but `registerAutoTypeFunction()` hasn't been called - at the very least a warning, but possibly an exception?